docs/reference/type-attribution.md

```diff
@@ -0,0 +1,385 @@
+
+All implementations of this interface provide a fully qualified name of the class (e.g., `"java.util.List"`). Also, except for `ShallowClass` and `Unknown`, the other implementations further provide access to the type's supertype, interfaces, methods, fields, and other metadata.
+
+When working with types it is common to check if a type is a `FullyQualified` instance before accessing its fully qualified name or other structural information.
+
+### JavaType implementations
+
+* **`JavaType.Class`** - Represents a fully-qualified class or interface with complete metadata:
+    * Fully qualified name (e.g., `"java.util.List"`)
+    * Supertype and interfaces
+    * Type parameters (e.g., `E` in `List<E>`)
+    * Fields (as `JavaType.Variable`)
+    * Methods (as `JavaType.Method`)
+    * Annotations
+    * Flags (public, static, final, etc.)
+
+* **`JavaType.Parameterized`** - Represents a parameterized type application:
+    * The raw type (a `FullyQualified`)
+    * Type arguments (e.g., `String` in `List<String>`)
+    * Example: `List<String>` vs. raw `List`
+
+* **`JavaType.ShallowClass`** - A lightweight `Class` variant containing only the fully qualified name:
+    * Used when a recipe needs to manually construct type information but only knows the fully qualified name
+    * Lacks the deep metadata (methods, fields, supertype) that a full `JavaType.Class` provides
+    * Useful as a fallback when full type information is unavailable
+    * **Best Practice**: Prefer obtaining complete type information from `JavaTemplate` or `JavaParser` when possible, but use `ShallowClass.build("com.example.MyClass")` if a more robust approach is unavailable.
+
+* **`JavaType.GenericTypeVariable`** - Represents type parameters and wildcards:
+    * Name (e.g., `"T"`, `"E"`, or `"?"` for wildcards)
+    * Bounds (e.g., `String` in `T extends String`)
+    * Variance (`INVARIANT`, `COVARIANT` for `? extends`, `CONTRAVARIANT` for `? super`)
+
+* **`JavaType.Array`** - Represents array types:
+    * Element type (which may itself be an array)
+    * Example: `String[]`, `int[][]`
+
+* **`JavaType.Primitive`** - Enum representing primitive types:
+    * Values: `Boolean`, `Byte`, `Char`, `Short`, `Int`, `Long`, `Float`, `Double`, `Void`, `String`, `Null`, `None`
+
+* **`JavaType.Method`** - Represents a method signature:
+    * Declaring type
+    * Method name
+    * Return type
+    * Parameter types and names
+    * Thrown exceptions
+    * Modifiers
+    * Type parameters declared by the method
+
+* **`JavaType.Variable`** - Represents a field or variable:
+    * Owner (the class or method declaring the variable)
+    * Variable type
+    * Variable name
+
+* **`JavaType.Unknown`** - Singleton sentinel value for unresolved types
+
+* **`JavaType.Intersection`** - Represents intersection types (e.g., `T extends A & B`)
```

Comment:
Minor correction: The intersection type is currently only used when doing casts like `(Function & Serializable)`, which is a really obscure Java feature.